applications/minotari_miner/src/run_miner.rs

```diff
@@ -330,12 +416,71 @@ async fn mining_cycle(
     let target_difficulty = miner_data.target_difficulty;
 
     debug!(target: LOG_TARGET, "Asking base node to assemble the MMR roots");
-    let block_result = node_conn.get_new_block(block_template).await?.into_inner();
-    let block = block_result.block.ok_or_else(|| err_empty("block"))?;
+    let block_result = base_node_client.get_new_block(block_template).await?.into_inner();
+    Ok(GetNewBlockResponse{
+        block: block_result.block.ok_or_else(|| err_empty("block"))?,
+        target_difficulty
+    })
+}
+
+async fn get_new_block_p2pool_node(sha_p2pool_client: &mut ShaP2PoolGrpcClient) -> Result<GetNewBlockResponse, MinerError> {
+    let block_result = sha_p2pool_client.get_new_block(GetNewBlockRequest::default()).await?.into_inner();
+    let new_block_result = block_result.block.ok_or_else(|| err_empty("block result"))?;
+    let block = new_block_result.block.ok_or_else(|| err_empty("block response"))?;
+    Ok(GetNewBlockResponse{
+        block,
+        target_difficulty: block_result.target_difficulty,
+    })
+}
+
+async fn submit_block(
+    config: &MinerConfig,
+    base_node_client: &mut BaseNodeGrpcClient,
+    sha_p2pool_client: Arc<Mutex<Option<ShaP2PoolGrpcClient>>>,
```

Comment:
I suspect you don't need a mutex.

```suggestion
    sha_p2pool_client: Option<&mut ShaP2PoolGrpcClient>,
```

Call with:
```rust
submit_block(base_node_client, sha_p2pool_client.as_mut(), ....)
```

Reply:
okay makes sense, just pushed